syftlib/src/main/java/org/openmined/syft/Syft.kt

```diff
@@ -132,35 +217,75 @@ class Syft internal constructor(
             this.workerId = workerId
     }
 
-    fun getSyftWorkerId() = workerId
+    private fun getNetworkState(workerId: String): Single<NetworkState> {
+        val networkState = NetworkState()
+        return getPing(workerId, networkState)
+                .andThen(getDownloadSpeed(workerId, networkState))
+                .andThen(getUploadSpeed(workerId, networkState))
+                .andThen(Single.just(networkState))
+    }
 
-    private fun getPing() = "1"
-    private fun getDownloadSpeed() = "1000"
-    private fun getUploadSpeed() = "1000"
+    private fun getPing(workerId: String, networkState: NetworkState): Completable {
+        val start = System.currentTimeMillis()
+        return getDownloader().checkPing(
+            workerId = workerId,
+            random = Random().ints(128).toString()
+        )
+                .flatMapCompletable { response ->
+                    if (response.code() == 200 && response.body()?.error == null) {
+                        networkState.ping = (System.currentTimeMillis() - start).toString()
+                        Log.d(TAG, "Ping is ${networkState.ping} ms")
+                        Completable.complete()
+                    } else
+                        Completable.error(NetworkErrorException("unable to get ping"))
 
-    private fun handleCycleReject(responseData: CycleResponseData.CycleReject) {
-        var jobId = SyftJob.JobID(responseData.modelName)
-        val job = workerJobs.getOrElse(jobId, {
-            jobId = SyftJob.JobID(responseData.modelName)
-            workerJobs.getValue(jobId)
-        })
-        job.cycleRejected(responseData)
+                }
+                .compose(networkingSchedulers.applyCompletableSchedulers())
     }
 
-    private fun handleCycleAccept(responseData: CycleResponseData.CycleAccept) {
-        val jobId = SyftJob.JobID(responseData.modelName)
-        val job = workerJobs.getOrElse(jobId, {
-            //todo change this when pygrid updates
-            workerJobs.getValue(
-                SyftJob.JobID(
-                    responseData.modelName,
-                    responseData.clientConfig.modelVersion
-                )
-            )
-        })
-        job.setJobArguments(responseData)
-        job.downloadData()
-    }
+    private fun getDownloadSpeed(workerId: String, networkState: NetworkState) = getDownloader()
```

Comment:
These functions should be in a separate class. Syft is doing too many different things. It should delegate.
A combination of repository/data sources should help with it.

As a side note: I guess the idea of having a desktop Kotlin version is not going anywhere anymore. If that's not yet the case, it would be good to have these classes in a separate module. Otherwise, just in a package.

Reply:
Am thinking of sending these methods into network state eventually. Network state is essentially the data source. Am still in the process of figuring out how to send worker specific details. I can send the workerId and just call these functions.